hub/core/tensor.py

```diff
@@ -115,33 +123,60 @@ def read_samples_from_tensor(
     key: str,
     storage: StorageProvider,
     index: Index = Index(),
-) -> np.ndarray:
+    aslist: bool = False,
+) -> Union[np.ndarray, List[np.ndarray]]:
     """Read (and unpack) samples from a tensor as an np.ndarray.
 
     Args:
         key (str): Key for where the chunks, index_map, and meta are located in `storage` relative to it's root.
         storage (StorageProvider): StorageProvider for reading the chunks, index_map, and meta.
         index (Index): Index that represents which samples to read.
+        aslist (bool): If True, a list of np.ndarrays will be returned. Helpful for dynamic tensors.
+            If False, a single np.ndarray will be returned unless the samples are dynamically shaped, in which case
+            an error is raised.
+
+    Raises:
+        DynamicTensorNumpyError: If reading a dynamically-shaped array without `aslist=True`.
+        NotImplementedError: Empty samples (shape contains 0) not implemented.
 
     Returns:
         np.ndarray: Array containing the sample(s) in the `array_slice` slice.
     """
 
     meta = read_tensor_meta(key, storage)
     index_map = read_index_map(key, storage)
+    index_entries = index_map[index.to_slice()]
+
+    dtype = meta["dtype"]
 
     # TODO: read samples in parallel
     samples = []
-    for index_entry in index_map[index.to_slice()]:
-        array = sample_from_index_entry(key, storage, index_entry, meta["dtype"])
-        samples.append(array)
+    is_fixed_shape = True
+    for i, index_entry in enumerate(index_entries):
+        shape = index_entry["shape"]
+
+        # check if all samples are the same shape
+        last_shape = index_entries[i - 1]["shape"]
+        if is_fixed_shape and i > 0 and shape != last_shape:
+            is_fixed_shape = False
 
-    array = np.array(samples)
+        if is_shape_empty(shape):
+            # TODO: implement support for 0s in shape and update docstring
+            raise NotImplementedError("0s in shapes are not supported yet.")
+
+        array = sample_from_index_entry(key, storage, index_entry, dtype)
+        samples.append(array)
 
     if isinstance(index.item, int):
-        array = array.squeeze(axis=0)
+        return samples[0]
+
+    if aslist:
+        return samples
+
+    if not is_fixed_shape:
```

Comment:
Not able to follow the use of `is_fixed_shape`. Can we raise the error as soon as we find this and not wait for samples to be added and then raise the error?
-> fail fast

Reply:
good call!